File: python/three_sums.py
from typing import List


def ThreeSums(numbers: List[int]) -> List[List[int]]:
    numbers.sort()
    res = []

    for index, number in enumerate(numbers):
        if index > 0 and number == numbers[index-1]:
            continue

        left, right = index+1, len(numbers)-1

        while left < right:
            three_sum = number + numbers[left] + numbers[right]

            if three_sum > 0:
                right -= 1
            elif three_sum < 0:
                left += 1
            else:
                res.append([number, numbers[left], numbers[right]])
                left += 1
                while numbers[left] == numbers[left-1] and left < right:
                    left += 1
    return res

File: go/three_sums.go
package main

import "slices"

func ThreeSums(numbers []int) [][]int {
	slices.Sort(numbers)
	res := [][]int{}

	for index, number := range numbers {
		if index > 0 && number == numbers[index-1] {
			continue
		}

		left, right := index+1, len(numbers)-1

		for left < right {
			three_sums := number + numbers[left] + numbers[right]

			if three_sums > 0 {
				right -= 1
			} else if three_sums < 0 {
				left += 1
			} else {
				res = append(res, []int{number, numbers[left], numbers[right]})
				left += 1
				for numbers[left] == numbers[left-1] && left < right {
					left += 1
				}
			}
		}
	}
	return res
}
